src/utils/first_follow.cpp

```diff
@@ -0,0 +1,267 @@
+  for (const auto &production : augmented_grammar) {
+    for (const auto &rule : production.GetRules()) {
+      for (const auto &symbol : rule.GetEntities()) {
+        // if symbol is not present as a parent in grammar, then it is a terminal
+        if (!SearchParent(augmented_grammar, symbol)) {
+          nullables[symbol] = false;
+        }
+      }
+    }
+  }
+
+  // EPSILON is nullable by defination
+  if (nullables.find(std::string(grammar::EPSILON)) != nullables.end()) {
+    nullables[std::string(grammar::EPSILON)] = true;
+  }
+
+  // recursively compute for all non terminals
+  std::function<bool(const std::string &, std::vector<std::string> &)> calc_recursive;
+  // lambda function that returns boolean value
+  calc_recursive = [&](const std::string &key, std::vector<std::string> &path) {
+    if (find(path.begin(), path.end(), key) != path.end()) {
+      return false;
+    }
+
+    // marking current symbol as visited
+    path.push_back(key);
+
+    // registering non-terminal in FOLLOW map if encountered for the first time
+    if (nullables.find(key) != nullables.end()) {
+      path.pop_back();
+      return nullables[key];
+    }
+
+    for (const auto &rule : GetRulesForParent(augmented_grammar, key)) {
+      std::vector<std::string>::const_iterator symbol_itr;
+      for (symbol_itr = rule.GetEntities().begin(); symbol_itr != rule.GetEntities().end(); symbol_itr++) {
+        // exit loop if non-nullable symbol is encountered
+        if (!calc_recursive(*symbol_itr, path)) {
+          break;
+        }
+      }
+
+      // for production A -> X Y Z, A is nullable iff X, Y, Z all are nullable
+      if (symbol_itr == rule.GetEntities().end()) {
+        nullables[key] = true;
+        path.pop_back();
+        return true;
+      }
+    }
+
+    // by default, a symbol is considered non-nullable
+    nullables[key] = false;
+    path.pop_back();
+    return false;
+  };
+
+  for (const auto &production : augmented_grammar) {
+    auto path = std::vector<std::string>(0);
+    calc_recursive(production.GetParent(), path);
+  }
+
+  return nullables;
+}
+
+std::unordered_map<std::string, std::vector<std::string>> CalcFirsts(
+    const grammar::Productions &augmented_grammar, const std::unordered_map<std::string, bool> &nullables) {
+  std::unordered_map<std::string, std::vector<std::string>> firsts;
+  // finished -> used to check if any new symbols are added to any FIRST(non-terminal) in a particular iteration
+  // useful in case of cycles in productions
+  bool finished = false;
+  for (const auto &production : augmented_grammar) {
+    for (const auto &rules : production.GetRules()) {
+      for (const auto &symbol : rules.GetEntities()) {
+        if (!SearchParent(augmented_grammar, symbol)) {
+          // if symbol is not present as a parent in grammar, then it is a terminal
+          // if X is a terminal, then FIRST(X) = {X}
+          firsts.insert(make_pair(symbol, std::vector<std::string>(1, symbol)));
```

Comment:
maintain consistency

Reply:
what consistency?

Reply:
[] ok

Reply:
Please don't use insert in case of `unordered_map` , becomes really unreadable.

Reply:
What does `std::vector<std::string>(1, symbol))` mean here.

Reply:
A better way would be 
```
firsts[symbol] = {symbol};
```

Reply:
exactly